File: poc/karaoke-sync-poc/src/parser.py
"""XML → structured data conversion for DAM scoring records.

Responsibilities:
- Extract typed fields from a <scoring> XML element
- Convert raw XML to JSON (for storage in scores.raw_xml)
- Compute MIDI note values from DAM's existing representation
- Normalize datetime strings
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

import structlog
import xmltodict
from lxml import etree
from pydantic import BaseModel, Field, field_validator

logger = structlog.get_logger(__name__)

JST = timezone.utc  # DAM returns UTC timestamps; adjust if needed later


# ---------------------------------------------------------------
# Pydantic models for structured score data
# ---------------------------------------------------------------


class ParsedScore(BaseModel):
    """Typed representation of a single scoring record.

    Field names mirror the `scores` table in schema.sql.
    """

    dam_scoring_id: str
    scoring_type: str  # ai / ai_heart / dxg / dx / other

    sung_at: datetime

    # Song identification
    song_title: str
    song_artist: str
    request_no: str | None = None
    dam_contents_id: str | None = None

    # Main score
    total_score: float

    # Radar (5 axes)
    pitch_score: float | None = None
    stability_score: float | None = None
    expression_score: float | None = None
    vibrato_longtone_score: float | None = None
    rhythm_score: float | None = None

    # Ai-specific
    ai_bonus: float | None = None

    # Performance context
    key_control: int = 0
    tempo_control: int | None = None
    guide_melody: bool | None = None

    # Range (MIDI note numbers)
    singing_range_lowest: int | None = None
    singing_range_highest: int | None = None
    vocal_range_lowest: int | None = None
    vocal_range_highest: int | None = None

    # 24-interval pitch data (only present when detailFlg=1)
    pitch_intervals: list[int] | None = None

    # Full XML as JSON for raw_xml column
    raw_xml: dict[str, Any]

    @field_validator("total_score", "pitch_score", "stability_score", "expression_score",
                     "vibrato_longtone_score", "rhythm_score", "ai_bonus", mode="before")
    @classmethod
    def _parse_float(cls, v: Any) -> float | None:
        if v is None or v == "":
            return None
        try:
            return float(v)
        except (TypeError, ValueError):
            return None

    @field_validator("key_control", "tempo_control", mode="before")
    @classmethod
    def _parse_int(cls, v: Any) -> int | None:
        if v is None or v == "":
            return None
        try:
            return int(v)
        except (TypeError, ValueError):
            return None

    @field_validator("singing_range_lowest", "singing_range_highest",
                     "vocal_range_lowest", "vocal_range_highest", mode="before")
    @classmethod
    def _parse_midi(cls, v: Any) -> int | None:
        if v is None or v == "":
            return None
        try:
            n = int(v)
            # Clamp to valid MIDI range (21-108 = piano range)
            if 21 <= n <= 108:
                return n
            return None
        except (TypeError, ValueError):
            return None


# ---------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------


def _get_attr(el: etree._Element, name: str) -> str | None:
    v = el.get(name)
    return v if v else None


def _get_child_text(el: etree._Element, tag: str) -> str | None:
    child = el.find(tag)
    if child is None or child.text is None:
        return None
    text = child.text.strip()
    return text if text else None


def _parse_datetime(s: str | None) -> datetime | None:
    """Parse DAM's datetime format.

    DAM typically returns: "2026-04-18 11:36:42" (no timezone)
    We treat it as JST and store as UTC in DB via timestamptz.
    """
    if not s:
        return None
    for fmt in ("%Y-%m-%d %H:%M:%S", "%Y-%m-%dT%H:%M:%S", "%Y%m%d%H%M%S"):
        try:
            return datetime.strptime(s.strip(), fmt).replace(tzinfo=timezone.utc)
        except ValueError:
            continue
    logger.warning("datetime_parse_failed", value=s)
    return None


def _infer_scoring_type(el: etree._Element) -> str:
    """Infer the scoring machine type from XML attributes.

    DAM encodes this in various ways; normalize to our ENUM values.
    """
    # Common attribute names seen in responses
    for attr in ("scoringType", "type", "machineType"):
        val = el.get(attr)
        if val:
            v = val.lower()
            if "heart" in v:
                return "ai_heart"
            if "dxg" in v or "dx-g" in v or "dx_g" in v:
                return "dxg"
            if "dx" in v:
                return "dx"
            if "ai" in v:
                return "ai"
    # Default: assume Ai (this endpoint is scoring-Ai specific)
    return "ai"


def _parse_pitch_intervals(el: etree._Element) -> list[int] | None:
    """Extract the 24-section pitch scores from intervalGraphPointsSection01..24 attrs.

    Live DAM response (detailFlg=1) exposes each section as a separate attribute:
        intervalGraphPointsSection01="71" ... intervalGraphPointsSection24="35"
    """
    values: list[int] = []
    for i in range(1, 25):
        raw = el.get(f"intervalGraphPointsSection{i:02d}")
        if raw is None or raw == "":
            return None
        try:
            values.append(int(raw))
        except ValueError:
            return None
    return values if len(values) == 24 else None


# ---------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------


def _scaled(v: str | None, divisor: float) -> float | None:
    if v is None or v == "":
        return None
    try:
        return float(v) / divisor
    except (TypeError, ValueError):
        return None


def parse_scoring_element(el: etree._Element) -> ParsedScore:
    """Parse one <scoring> element into a ParsedScore.

    Attribute mapping is based on live DAM responses (verified 2026-04).
    The API is undocumented so we parse defensively (everything optional
    except the core identifiers and score).
    """
    scoring_ai_id = _get_attr(el, "scoringAiId")
    if not scoring_ai_id:
        raise ValueError("scoring element has no scoringAiId")

    # Sung time: `scoringDateTime` is YYYYMMDDHHMMSS
    sung_at_raw = (
        _get_attr(el, "scoringDateTime")
        or _get_attr(el, "scoringDate")
        or _get_child_text(el, "scoringDateTime")
        or _get_child_text(el, "scoringDate")
    )
    sung_at = _parse_datetime(sung_at_raw)
    if sung_at is None:
        raise ValueError(f"scoring {scoring_ai_id}: could not parse sung_at from '{sung_at_raw}'")

    # Song info
    song_title = (
        _get_attr(el, "contentsName")
        or _get_child_text(el, "contentsName")
        or _get_attr(el, "songName")
        or "(unknown)"
    )
    song_artist = (
        _get_attr(el, "artistName")
        or _get_child_text(el, "artistName")
        or "(unknown)"
    )

    # Total score is stored as the element text content (integer * 1000).
    # e.g. <scoring ...>90298</scoring> → 90.298
    # Fallback to legacy attribute names if format changes.
    total_raw = (el.text or "").strip() or None
    total_score = _scaled(total_raw, 1000.0)
    if total_score is None:
        # Legacy fallback
        total_score = _scaled(
            _get_attr(el, "scoringResult") or _get_attr(el, "totalScore"), 1000.0
        )

    # Serialize raw XML to JSON for storage
    raw_xml_str = etree.tostring(el, encoding="unicode")
    raw_xml_json = xmltodict.parse(raw_xml_str)

    return ParsedScore(
        dam_scoring_id=scoring_ai_id,
        scoring_type=_infer_scoring_type(el),
        sung_at=sung_at,
        song_title=song_title,
        song_artist=song_artist,
        request_no=_get_attr(el, "requestNo") or _get_child_text(el, "requestNo"),
        dam_contents_id=_get_attr(el, "contentsId") or _get_child_text(el, "contentsId"),
        total_score=total_score,  # validated in Pydantic
        # Radar: raw attrs are 0-100 integers, no scaling.
        pitch_score=_get_attr(el, "radarChartPitch"),
        stability_score=_get_attr(el, "radarChartStability"),
        expression_score=_get_attr(el, "radarChartExpressive"),
        vibrato_longtone_score=_get_attr(el, "radarChartVibratoLongtone"),
        rhythm_score=_get_attr(el, "radarChartRhythm"),
        # aiSensitivityBonus is integer * 1000 (e.g. 3723 → 3.723)
        ai_bonus=_scaled(_get_attr(el, "aiSensitivityBonus"), 1000.0),
        key_control=_get_attr(el, "lastPerformKey") or 0,
        tempo_control=_get_attr(el, "tempoControl"),
        guide_melody=(_get_attr(el, "guideMelody") == "1") if _get_attr(el, "guideMelody") else None,
        singing_range_lowest=_get_attr(el, "singingRangeLowest"),
        singing_range_highest=_get_attr(el, "singingRangeHighest"),
        vocal_range_lowest=_get_attr(el, "vocalRangeLowest"),
        vocal_range_highest=_get_attr(el, "vocalRangeHighest"),
        pitch_intervals=_parse_pitch_intervals(el),
        raw_xml=raw_xml_json,
    )


# ---------------------------------------------------------------
# MIDI utilities (exposed for the app)
# ---------------------------------------------------------------


NOTE_NAMES = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]


def midi_to_note_name(midi: int) -> str:
    """Convert MIDI note number to note name (e.g. 60 -> 'C4')."""
    if not (21 <= midi <= 108):
        return f"?({midi})"
    octave = (midi // 12) - 1
    name = NOTE_NAMES[midi % 12]
    return f"{name}{octave}"

File: supabase/functions/sync-scores/parser.ts
/**
 * XML <scoring> → ParsedScore conversion. Mirrors poc/karaoke-sync-poc/src/parser.py
 * attribute mappings that were empirically verified in Phase 1 (see
 * docs/data-model.md and the parser.py fixes commit).
 */

export type ScoringTypeT =
  | "ai"
  | "ai_heart"
  | "dxg"
  | "dx"
  | "other";

export type ParsedScore = {
  dam_scoring_id: string;
  scoring_type: ScoringTypeT;
  sung_at: Date;
  song_title: string;
  song_artist: string;
  request_no: string | null;
  dam_contents_id: string | null;
  total_score: number;
  pitch_score: number | null;
  stability_score: number | null;
  expression_score: number | null;
  vibrato_longtone_score: number | null;
  rhythm_score: number | null;
  ai_bonus: number | null;
  key_control: number;
  tempo_control: number | null;
  guide_melody: boolean | null;
  singing_range_lowest: number | null;
  singing_range_highest: number | null;
  vocal_range_lowest: number | null;
  vocal_range_highest: number | null;
  pitch_intervals: number[] | null;
  raw_xml: unknown; // xml-as-json for archival in scores.raw_xml (JSONB)
};

// fast-xml-parser emits attributes under this prefix when configured below.
export const ATTR_PREFIX = "@_";
export const TEXT_NODE = "#text";

function attr(el: Record<string, unknown>, name: string): string | null {
  const v = el[`${ATTR_PREFIX}${name}`];
  if (v === undefined || v === null || v === "") return null;
  return String(v);
}

function scaled(v: string | null, divisor: number): number | null {
  if (v === null) return null;
  const n = Number(v);
  if (!Number.isFinite(n)) return null;
  return n / divisor;
}

function midi(v: string | null): number | null {
  if (v === null) return null;
  const n = Number(v);
  if (!Number.isFinite(n)) return null;
  return n >= 21 && n <= 108 ? Math.trunc(n) : null;
}

function integer(v: string | null): number | null {
  if (v === null) return null;
  const n = Number(v);
  return Number.isFinite(n) ? Math.trunc(n) : null;
}

function parseDamDateTime(v: string | null): Date | null {
  if (!v) return null;
  // YYYYMMDDHHMMSS
  const m14 = /^(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})$/.exec(v);
  if (m14) {
    const [, y, mo, d, h, mi, s] = m14;
    // DAM returns local JST but server stores as UTC. Treat as UTC for consistency
    // with the Python PoC (which did the same).
    return new Date(Date.UTC(+y, +mo - 1, +d, +h, +mi, +s));
  }
  // Fallbacks: "YYYY-MM-DD HH:MM:SS", ISO, etc.
  const d = new Date(v.replace(" ", "T") + "Z");
  return Number.isNaN(d.getTime()) ? null : d;
}

function inferScoringType(el: Record<string, unknown>): ScoringTypeT {
  for (const a of ["scoringType", "type", "machineType"]) {
    const v = attr(el, a);
    if (!v) continue;
    const s = v.toLowerCase();
    if (s.includes("heart")) return "ai_heart";
    if (s.includes("dxg") || s.includes("dx-g") || s.includes("dx_g")) return "dxg";
    if (s.includes("dx")) return "dx";
    if (s.includes("ai")) return "ai";
  }
  return "ai";
}

function parsePitchIntervals(el: Record<string, unknown>): number[] | null {
  const values: number[] = [];
  for (let i = 1; i <= 24; i++) {
    const key = `intervalGraphPointsSection${i.toString().padStart(2, "0")}`;
    const raw = attr(el, key);
    if (raw === null) return null;
    const n = Number(raw);
    if (!Number.isFinite(n)) return null;
    values.push(Math.trunc(n));
  }
  return values.length === 24 ? values : null;
}

/**
 * Parse one <scoring> element (as yielded by fast-xml-parser) into a
 * ParsedScore. Returns null and logs when required fields are missing.
 */
export function parseScoringElement(
  el: Record<string, unknown>,
): ParsedScore {
  const scoring_ai_id = attr(el, "scoringAiId");
  if (!scoring_ai_id) {
    throw new Error("scoring element has no scoringAiId");
  }

  const sungRaw =
    attr(el, "scoringDateTime") ??
    attr(el, "scoringDate");
  const sung_at = parseDamDateTime(sungRaw);
  if (!sung_at) {
    throw new Error(
      `scoring ${scoring_ai_id}: could not parse sung_at from '${sungRaw}'`,
    );
  }

  const song_title =
    attr(el, "contentsName") ??
    attr(el, "songName") ??
    "(unknown)";
  const song_artist = attr(el, "artistName") ?? "(unknown)";

  // Total score lives in the element's text content. fast-xml-parser with
  // textNodeName:"#text" exposes it under TEXT_NODE.
  const totalRaw = el[TEXT_NODE];
  const total_score =
    scaled(
      totalRaw !== undefined && totalRaw !== null ? String(totalRaw) : null,
      1000,
    ) ??
    scaled(attr(el, "scoringResult") ?? attr(el, "totalScore"), 1000);
  if (total_score === null) {
    throw new Error(`scoring ${scoring_ai_id}: missing total_score`);
  }

  const guideMelodyRaw = attr(el, "guideMelody");

  return {
    dam_scoring_id: scoring_ai_id,
    scoring_type: inferScoringType(el),
    sung_at,
    song_title,
    song_artist,
    request_no: attr(el, "requestNo"),
    dam_contents_id: attr(el, "contentsId"),
    total_score,
    pitch_score: integer(attr(el, "radarChartPitch")),
    stability_score: integer(attr(el, "radarChartStability")),
    expression_score: integer(attr(el, "radarChartExpressive")),
    vibrato_longtone_score: integer(attr(el, "radarChartVibratoLongtone")),
    rhythm_score: integer(attr(el, "radarChartRhythm")),
    ai_bonus: scaled(attr(el, "aiSensitivityBonus"), 1000),
    key_control: integer(attr(el, "lastPerformKey")) ?? 0,
    tempo_control: integer(attr(el, "tempoControl")),
    guide_melody: guideMelodyRaw === null ? null : guideMelodyRaw === "1",
    singing_range_lowest: midi(attr(el, "singingRangeLowest")),
    singing_range_highest: midi(attr(el, "singingRangeHighest")),
    vocal_range_lowest: midi(attr(el, "vocalRangeLowest")),
    vocal_range_highest: midi(attr(el, "vocalRangeHighest")),
    pitch_intervals: parsePitchIntervals(el),
    raw_xml: el,
  };
}
